src/state/state-manager.js

```diff
@@ -68,87 +69,49 @@ export default class StateManager {
     },
     [StateType.LOADING]: {
-      [Html5EventType.LOADED_METADATA]: () => {
-        this._updateState(StateType.PAUSED);
-        this._dispatchEvent();
-      },
-      [Html5EventType.ERROR]: () => {
-        this._updateState(StateType.IDLE);
-        this._dispatchEvent();
-      },
+      [Html5EventType.LOADED_METADATA]: () => this._updateState(StateType.PAUSED),
+      [Html5EventType.ERROR]: () => this._updateState(StateType.IDLE),
       [Html5EventType.SEEKED]: () => {
         if (this._prevState && this._prevState.type === StateType.PLAYING) {
           this._updateState(StateType.PLAYING);
-          this._dispatchEvent();
         }
       }
     },
     [StateType.PAUSED]: {
-      [Html5EventType.PLAY]: () => {
-        this._updateState(StateType.PLAYING);
-        this._dispatchEvent();
-      },
-      [Html5EventType.PLAYING]: () => {
-        this._updateState(StateType.PLAYING);
-        this._dispatchEvent();
-      },
-      [Html5EventType.ENDED]: () => {
-        this._updateState(StateType.IDLE);
-        this._dispatchEvent();
-      }
+      [Html5EventType.PLAY]: () => this._updateState(StateType.PLAYING),
+      [Html5EventType.PLAYING]: () => this._updateState(StateType.PLAYING),
+      [Html5EventType.ENDED]: () => this._updateState(StateType.IDLE)
     },
     [StateType.PLAYING]: {
-      [Html5EventType.PAUSE]: () => {
-        this._updateState(StateType.PAUSED);
-        this._dispatchEvent();
-      },
+      [Html5EventType.PAUSE]: () => this._updateState(StateType.PAUSED),
       [Html5EventType.WAITING]: () => {
         if (this._player.seeking) {
           this._updateState(StateType.LOADING);
-          this._dispatchEvent();
         } else {
           this._updateState(StateType.BUFFERING);
           this._lastWaitingTime = this._player.currentTime;
-          this._dispatchEvent();
         }
       },
-      [Html5EventType.ENDED]: () => {
-        this._updateState(StateType.IDLE);
-        this._dispatchEvent();
-      },
-      [Html5EventType.ERROR]: () => {
-        this._updateState(StateType.IDLE);
-        this._dispatchEvent();
-      }
+      [Html5EventType.ENDED]: () => this._updateState(StateType.IDLE),
+      [Html5EventType.ERROR]: () => this._updateState(StateType.IDLE)
     },
     [StateType.BUFFERING]: {
-      [Html5EventType.PLAYING]: () => {
-        this._updateState(StateType.PLAYING);
-        this._dispatchEvent();
-      },
-      [Html5EventType.PAUSE]: () => {
-        this._updateState(StateType.PAUSED);
-        this._dispatchEvent();
-      },
+      [Html5EventType.PLAYING]: () => this._updateState(StateType.PLAYING),
+      [Html5EventType.PAUSE]: () => this._updateState(StateType.PAUSED),
       [Html5EventType.TIME_UPDATE]: () => {
-        if (this._player.currentTime !== this._lastWaitingTime && this._prevState && this._prevState.type === StateType.PLAYING) {
+        if (
+          Env.browser.name === 'IE' &&
```

Comment:
So actually this is the only change in logic, right? the rest seems to be refactoring

Reply:
Yes...
Just changing the event dispatching to be from `updateState` because you always want to do it and there's no need to split it outside and so many places.